### DISCUSSION TITLE
Improve the Process for Approving Posts Under a Tag that Requires Approval

### DISCUSSION THREAD

**Discussion by michael-tharrington** [forem/forem#16054] [Feature Requests]
## Summary

We need a better way for admin to approve posts under a tag that requires approval like [#shecoded](https://dev.to/t/shecoded). For tags that require approval, we don’t yet have a really great way for admin to see which articles we’ve approved and which articles we haven’t approved under that tag.

## Problem Statement & Context

We recently discussed DEV’s [Nevertheless, She Coded](https://dev.to/shecoded) event that’s coming up in March and how we could prepare in advance to make everything run as smoothly as possible.

During the meeting, we recognized an issue that has come up in past years and made it somewhat difficult to moderate the event. The issue revolves around our mechanism for approving [#shecoded](https://dev.to/t/shecoded) articles.

Let me first explain how tag approval works. Admin can edit a tag’s details (from `/admin/content_manager/tags/#/edit`) so that the tag requires approval. This means that any article published under the tag (e.g. #shecoded) must be approved by an admin before it appears on the tag’s landing page. For an admin to approve an article under the tag, the admin must visit the article’s admin page (i.e. `/admin/content_manager/articles/#`), check the box beside “Approved”, and click submit.

In the case of approving articles shared under #shecoded, we don’t yet have a really great way for admin to see which articles we’ve approved and which articles we haven’t approved. I’ve filmed [a loom video](https://www.loom.com/share/723b4ec1b5df4c8398478de2fb0cc97e) to show our current process for approving posts. In the video, I explain how we use the `/mod` view to filter to articles tagged with #shecoded and described some of the issues we face when trying to keep track of approved/unapproved posts.

## Potential Solutions

1. **We come up with new helpful filters for the `/admin/content_manager/articles` page.** One filter would allow admin to filter articles by tag (e.g. `/admin/content_manager/articles?tag=shecoded`) and another would allow us to see unapproved or approved posts under a tag (e.g. `/admin/content_manager/articles?tag=shecoded&approved=false`) Note: my filter writing may be off, but hope that at least helps convey what I’m thinking about. 

2. **Allow mods of a tag that requires approval to approve the tag via their mod panel.** As admins we’re already using `/mod/shecoded` to filter to #shecoded posts. If we could approve them from this interface, then we could more easily see where we’re at in the approval process. We could even loop in mods to help us with all of this! This would make the feature much more robust in my opinion.

Also wanted to note a bunch of other ideas that @ludwiczakpawel has thrown out in addition to the 2 above:

3. add an indicator in the /mod view table to tell if something is waiting for approval.. this could be either an extra column or just an icon in front of the post title (similar to what we do with “:wave:” for first time authors).

4. make unapproved articles somehow visible for mods on tag page (i.e. /t/shecoded).. i think this could be handled in couple of ways:
    - add a button somewhere saying something along the lines “8 posts waiting for approval” which would ultimately take you the mod to, for example, appropriate /mod page
    - utilize one of the sidebars to show list of unapproved (pending?) posts so they can quickly see such a list and go through it.
    - weird/wild one: add extra tab on top of the feed view, next to other tabs (“relevant, latest, top”) with “Mod” (or something like that) that would show you only posts that require mod action (e.g. approval)
    - display unapproved posts the same way as others (inside the feed) but with some special indicator AND action button “approve” next to unapproved post (probably similar to how we indicate “Pinned” post on home page)

5. add a notification on individual post view that such a post requires an approval and add a button to approve it, right from that view. obviously this would be only visible to mods/admins.

## Drawbacks

No major drawbacks in my mind. 

On idea #2, It's possible that some creators wouldn't want tag moderators to have the ability to approve posts, but then, I think if that's the case, they most likely wouldn't want the tag moderator to be a moderator for a tag that requires approval.

## (Optional) Additional Context

Nothing too much to add at this moment, but I'm tagging in @amywtlin who I know has thoughts on this!

**Comment by amywtlin** [comment]
A bit more context from a Product perspective: 

## In the long run

I envision this to be solved with the [guest writing scenario](https://docs.google.com/document/d/1mPd40UazOsa6gLoD4AXVpXaXN9GGIACLfhu8HCnF-rE/edit?pli=1#heading=h.9kkk98cvfvbs) in our Roles System project, where admins can create a content group where posts are only published upon approval. I also added a paragraph in the doc on related migration.

## In the near future

I agree with all the suggestions mentioned here. I want us to be intentional of what suggestions solve what problems, so I am recapping the problem statements I observe here:

1. As a mod/admin, I can't easily see which posts under the tag are to be approved in the mod center. I need to click into the mod/admin page of each one of the posts to see if I need to approve them.
2. As a mod/admin, I can't approve a post easily right where I am (in the video, it's the list view); I need to go to the mod/admin page of that post to approve it.

I also groomed the solutions mentioned above, and mapped them back to what they solve:
1. Solutions for problem 1:
   1. Add filters on post views: Those filters needs to be well thought of, we will have what filters do we need and can be helpful.
   1. add an indicator in the /mod view table to tell if something is waiting for approval.. this could be either an extra column or just an icon in front of the post title (similar to what we do with “:wave:” for first time authors).
   1. make unapproved articles somehow visible for mods on tag page (i.e. /t/shecoded).
   1. add a notification on individual post view that such a post requires an approval and add a button to approve it, right from that view. obviously this would be only visible to mods/admins.
1. Solutions for problem 2:
   1. Bulk-approve: Another thing I was thinking here was bulk actions. This will really make the job easier I guess.
   1. Bring approve/bulk-approve to mod post view: Such mod actions really makes sense in mod panel actually.
   
@anujbhavsar96 @ludwiczakpawel among the solutions mentioned (or any other new ideas that come to your mind), which one do you think would be most impactful? Which one(s) do you feel like could be Small Wins?

**Reply by ludwiczakpawel** [reply]
> @anujbhavsar96 @ludwiczakpawel among the solutions mentioned (or any other new ideas that come to your mind), which one do you think would be most impactful? Which one(s) do you feel like could be Small Wins?

I think it would be best if @michael-tharrington could tell us...

I think this one would be **low effort and high impact**:

> Add an indicator in the /mod view table to tell if something is waiting for approval.. this could be either an extra column or just an icon in front of the post title (similar to what we do with “👋” for first time authors).

...And this could also be **high impact** (probably **medium effort** though):

> Allow mods of a tag that requires approval to approve the tag via their mod panel

...And last but not least, but this might be (depending on the approach) **medium or high effort**:

> Come up with new helpful filters for the /admin/content_manager/articles page

**Reply by amywtlin** [reply]
I have given it a bit more thoughts..........

I think it makes the most sense for this set of features to be in `/content_manager`, assuming we don't get to the use case in Roles System that solves this problem soon enough. Admins can already approve posts from the list view, which we can also extend to bulk-approve. 

Extending the approval functionality to `/mod` might also help, but I feel like it is because that we can view posts under a specific tag there and it feels like a hack to me, not where it belongs. Unless we have a solid use case where we would involve mods in approving posts. 

@michael-tharrington How likely do you think we'd be inviting mods to help with this though? Can you give us a few examples where we could have done it in the past, if mods were able to approve posts?

Among all options, I think I want us to look into revamping `/content_manager` and include filtering with tags and with approval status the most, with the extensibility for more filters in the future in mind. Though admin overhaul is currently blocked by my lack of bandwidth to come back to it....... But I think this is the right way to go.

@JennieOcken How do you feel about this approach though?

**Reply by ludwiczakpawel** [reply]
> I feel like it is because that we can view posts under a specific tag there and it feels like a hack to me, not where it belongs

Why you think so? Approving posts is tightly connected to tags - whether a post must be approved or not is literally specified on a tag level. So I had quite the opposite feeling - that it does perfectly fit some tag-related view. 

And that in general felt like a "natural feature (permission?)" for **mods** to be able to approve posts.

But I might have misunderstood something here as well :DDD 

---

Btw. completely separate issue but I'm super curious - what happens with "rejects" posts? Do we have a flag like that?

**Reply by anujbhavsar96** [reply]
> Among all options, I think I want us to look into revamping /content_manager and include filtering with tags and with approval status the most, with the extensibility for more filters in the future in mind. 

I feel this is the right direction to start with. And it aligns with our future plans as well.

**Reply by amywtlin** [reply]
> Why you think so? Approving posts is tightly connected to tags - whether a post must be approved or not is literally specified on a tag level.

If you think about it - approving posts is one of the actions you do to a post, right? Supposedly, it makes sense for all posts related actions to be in the same place, such as deleting a post, pinning a post, etc. That's mainly where I am coming from.

Approving posts is tightly connected to tags because that's how we have set things up in the past. Going forward, to simplify the purpose and the use case of tags, ("[tags should be added by the author of a post, with the goal to increase the discoverability of the post](https://miro.com/app/board/o9J_l2tsfZo=/?moveToWidget=3074457365128894495&cot=14)") I think approving posts is exactly one of the use cases we should move away from tags. So I would prefer the interim approach to not get tangled up with tags too much, to facilitate the transition and make sure the new functionality would remain useful, no matter how we handle approving posts, with tags or later with content groups. To put it differently, I think `/content-manager` should be able to handle "approving posts" no matter it is under a tag or in a content group. Does this makes sense?

I am not saying we shouldn't add that functionality to /mod though - it makes total sense if we involve mods for this operations. I just feel we are more scratching the surface with this approach, instead of handling the root cause. I dig a bit more to make sure that's what we see ourselves doing a lot though, hence my question for @michael-tharrington on how likely we think we will involve mods in this operation.
